### DISCUSSION TITLE
Debug support

### DISCUSSION THREAD

**Discussion by dev-area** [polhenarejos/pico-hsm#31] [Q&A] [answered]
Hi
Is there any easy way to debug the code (printf/jtag/serial port)?
thanks
liran

**Comment by polhenarejos** [comment, marked as answer]
Firmware must be recompiled with DEBUG_APDU=1 and attach the pico's UART to your machine through a serial port. UART pins depend on your specific pico device but your vendor will provide a pinout diagram for sure. Then open a console and connect it to the serial port at 115200 bauds. All CCID in/out messages will appear in the screen.

Another quick way to debug (but more tedious) is using Wireshark to sniff all USB packets. It is much simpler but USB payloads contain more garbage. 

What do you want to debug specifically?

**Reply by dev-area** [reply]
Thanks
I need to add 3DES support (for an old mechanism I use and cant be changed to AES)
i succeeded to add special instructions to store and encrypt/decrypt  (using mdeblts i also had to configure)
it is working fine but i want to do it in the right way (using key gen commands etc) 
means i need to make changes inside your code and it cant be done without debug
the pico UART is great for me
where should i set the DEBUG_APDU ? (found many places)
thanks
liran

**Reply by polhenarejos** [reply]
In that case, reconfigure with `cmake .. -DDEBUG_APDU=1` and build it again.

Note that NIST deprecated 3DES in 2017 and it was removed from Mbedtls for being vulnerable. Despite 3DES primitives are still present (not ciphersuites), it is probable that will be removed in the future. In that case, you would need to revert your Mbedtls version to old versions, which may break part of Pico HSM code making it not compatible to future releases.

**Reply by dev-area** [reply]
i know 3DES is weak and deprecated but there is a mechanism in our eID that is based on it , to change it I need to replace all the citizens ID cards :)
thanks

**Reply by dev-area** [reply]
hi
just to summaries 
the UART solution works like a charm
added the -DDEBUG_APDU=1
connected the pico board using pins 21,22 and ground
using minicom 

thanks!!!

**Reply by polhenarejos** [reply]
Good to know it. Feel free to share any progress you do with 3DES.
